res/res_agi.c

```diff
@@ -1091,111 +1090,85 @@ static int add_silence(struct ast_channel *chan, struct ast_frame *f, struct ast
         long int ts_diff=0, f_ptime, last_seq=0;
 	int64_t gap_ms=0;
 	struct timeval s_tv = ast_channel_get_rec_start_time(chan);
+	unsigned int themssrc=0;
+	int ssrc_change=0;
+
+        /*! Check SSRC */
+        if ((f->themssrc != 0) && (f->themssrc != ast_channel_get_last_ssrc(chan, stream_no))){
+                ast_debug(1, "STREAM %d ==== SSRC changed (%u) to (%u)\n", stream_no, ast_channel_get_last_ssrc(chan, stream_no), f->themssrc);
+                ast_channel_set_last_ssrc(chan, f->themssrc, stream_no);
+                themssrc = f->themssrc;
+		ssrc_change=1;
+        } else if (f->themssrc == ast_channel_get_last_ssrc(chan, stream_no)) {
+                themssrc = ast_channel_get_last_ssrc(chan, stream_no);
+        }else {
+		//Discard the packet
+		return 0;
+	}
 
-        if ((stream_no == 1) && (ast_channel_get_s1_pkt_count(chan) == 0)){
-                /*! Save the ts and sequence number for the next check */
-                //ast_channel_set_s1_last_ts(chan, f->ts);
-                //ast_channel_set_s1_last_seq(chan, f->seqno);
-                ast_channel_set_s1_pkt_count(chan);
-		f_ptime=ast_channel_get_s1_ptime(chan);
-                if(f_ptime < 5)
-                        f_ptime=20;
+	/*! ptime for the stream */
+	f_ptime=ast_channel_get_ptime(chan, stream_no);
+	if(f_ptime < 5)
+		f_ptime=20;
+
+	/*! First - Check if it is the 1st packet for the stream and fix the initial delay of streams */
+        if ((ast_channel_get_pkt_count(chan, stream_no) == 0) || (ssrc_change == 1)){
```

Comment:
If we re-order this check, we could potentially reduce one function call in most cases. Like,
`if ((ssrc_change || !ast_channel_get_pkt_count(chan, stream_no))`